[Merged by Bors] - Add test_logger as feature to logging

winksaville opened the pull request:
## Issue Addressed

Fix #2585

## Proposed Changes

Provide a canonical version of test_logger that can be used
throughout lighthouse.

## Additional Info

This allows tests to conditionally emit logging data by adding
test_logger as the default logger. And then when executing
`cargo test --features logging/test_logger` log output
will be visible:

  wink@3900x:~/lighthouse/common/logging/tests/test-feature-test_logger (Add-test_logger-as-feature-to-logging)
  $ cargo test --features logging/test_logger
      Finished test [unoptimized + debuginfo] target(s) in 0.02s
       Running unittests (target/debug/deps/test_logger-e20115db6a5e3714)

  running 1 test
  Sep 10 12:53:45.212 INFO hi, module: test_logger:8
  test tests::test_fn_with_logging ... ok

  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Doc-tests test-logger

  running 0 tests

  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

Or, in normal scenarios where logging isn't needed, executing
`cargo test` the log output will not be visible:

  wink@3900x:~/lighthouse/common/logging/tests/test-feature-test_logger (Add-test_logger-as-feature-to-logging)
  $ cargo test
      Finished test [unoptimized + debuginfo] target(s) in 0.02s
       Running unittests (target/debug/deps/test_logger-02e02f8d41e8cf8a)

  running 1 test
  test tests::test_fn_with_logging ... ok

  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Doc-tests test-logger

  running 0 tests

  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

winksaville committed:
Add test_logger as feature to logging

Provide a canonical version of test_logger that can be used
throughout lighthouse.

This allows tests to conditionally emit logging data by adding
test_logger as the default logger. And then when executing
`cargo test --features logging/test_logger` log output
will be visible:

  wink@3900x:~/lighthouse/common/logging/tests/test-feature-test_logger (Add-test_logger-as-feature-to-logging)
  $ cargo test --features logging/test_logger
      Finished test [unoptimized + debuginfo] target(s) in 0.02s
       Running unittests (target/debug/deps/test_logger-e20115db6a5e3714)

  running 1 test
  Sep 10 12:53:45.212 INFO hi, module: test_logger:8
  test tests::test_fn_with_logging ... ok

  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Doc-tests test-logger

  running 0 tests

  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

Or, in normal scenarios where logging isn't needed, executing
`cargo test` the log output will not be visible:

  wink@3900x:~/lighthouse/common/logging/tests/test-feature-test_logger (Add-test_logger-as-feature-to-logging)
  $ cargo test
      Finished test [unoptimized + debuginfo] target(s) in 0.02s
       Running unittests (target/debug/deps/test_logger-02e02f8d41e8cf8a)

  running 1 test
  test tests::test_fn_with_logging ... ok

  test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Doc-tests test-logger

  running 0 tests

  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

Fix #2585

winksaville commented on common/logging/Cargo.toml:
```diff
@@ -4,8 +4,12 @@ version = "0.2.0"
 slog-term = "2.6.0"
 lighthouse_metrics = { path = "../lighthouse_metrics" }
 lazy_static = "1.4.0"
+sloggers = "1.0.1"
```
Change to 2.0.2

winksaville commented on common/logging/tests/test.rs:
```diff
@@ -0,0 +1,32 @@
+    println!("test_test_logger_with_feature_test_logger output: {:#?}", output);
+
+    // Assert output data DOES contain "INFO hi, "
+    let data = String::from_utf8(output.unwrap().stdout).unwrap();
```
Change `.stdout` to `.stderr`.

winksaville commented on common/logging/tests/test.rs:
```diff
@@ -0,0 +1,32 @@
+    println!("test_test_logger_no_features output: {:#?}", output);
+
+    // Assert output data DOES contain "INFO hi, "
+    let data = String::from_utf8(output.unwrap().stdout).unwrap();
```
Change .stdout to .stderr

winksaville reviewed (commented):
If #2588 is merged these additional changes are needed.

winksaville committed:
Change slasher to use the common test_logger.

winksaville committed:
Change beacon_node/beacon_chain to use the common test_logger.

winksaville commented:
This is ready for review, it has been rebased on top of latest unstable and I've resolved necessary changes needed since #2588 was merged into unstable. Also, I've changed slasher and beacon_node/beacon_chain to use logging::test_logger as two separate commits on top of the revised original.

winksaville commented:
Change the test_test_logger tests to only run in debug builds as they are failing in release builds. But, interesting enough on my computer both work. I've no idea why.

winksaville commented:
Back to "draft" I'm going to try to make the tests pass the release-tests-windows and also add some documentation.

winksaville committed:
Fix issue with test-release-windows failing

The problem was the Command needed to be windows specific so I created
run_cmd to which used `if cfg!(target_os = "windows")` do the right
thing for windows and other platforms.

winksaville commented:
I this first pass is ready for review, txs!

winksaville commented:
Cleanup doc changes in setup.md and the previous CI tests failed in `eth1-simulator-ubuntu` for no apparent reason. I looked back in Actions history and [the same error has occurred previousl](https://github.com/sigp/lighthouse/runs/3615697961?check_suite_focus=true). So we'll see how this run goes.

michaelsproul commented on common/logging/src/lib.rs:
````diff
@@ -157,3 +158,70 @@ impl<'a> slog_term::RecordDecorator for AlignedRecordDecorator<'a> {
+/// Sep 19 18:39:20.537 INFO Beacon chain initialized, head_slot: 0, head_block: 0xa739…1b22, head_state: 0xac1c…eab6, module: beacon_chain::builder:649
+/// Sep 19 18:39:20.538 INFO Saved beacon chain to disk, module: beacon_chain::beacon_chain:3608
+/// test validator_pubkey_cache::test::basic_operation ... ok
+/// ```
````
I think this comment is excessively long. I think it suffices to say something like

```
By default no logs will be printed, but they can be enabled via the `test_logger` feature.

This feature can be enabled for any dependent crate by passing `--features logging/test_logger`, e.g.

cargo test -p beacon_chain --features logging/test_logger
```

winksaville replied:
Done

michaelsproul commented on book/src/setup.md:
```diff
@@ -36,6 +35,120 @@ you can run them locally and avoid CI failures:
+
+#### test_logger
+
+_I suspect this is not the right place for this documtation, but I think it
+would be good if it was discoverable and "all" developers will see it here_
```
I think this a fine spot for this content

michaelsproul replied:
We should delete this paragraph before merging

winksaville replied:
Done

michaelsproul commented on book/src/setup.md:
```diff
@@ -8,21 +8,20 @@ Lighthouse. This will install Lighthouse to your `PATH`, which is not
+  then the directory `~/eth2-lighthouse` is **\<rr\>**.
 
+## Using `make`
 We also wrap some of these commands and expose them via the `Makefile` in the
```
```suggestion
Commands to run the test suite are available via the `Makefile` in the
```

winksaville replied:
Done

michaelsproul commented on book/src/setup.md:
````diff
@@ -8,21 +8,20 @@ Lighthouse. This will install Lighthouse to your `PATH`, which is not
-cargo test
-```
+## Nomenclature
+- **\<rr\>** Repo Root. Example, if you `git clone https://github.com/sigp/lighthouse.git ~/eth2-lighthouse`
````
This looks a little bit like a stray HTML tag in the rendered docs. I like the repo root concept but I think it's usually clear from context that e.g. `beacon_node/beacon_chain` means to start from the repo root. This is the convention used by the Rust compiler itself, and I think is likely to be widely understood.

winksaville replied:
OK

michaelsproul commented on book/src/setup.md:
```diff
@@ -36,6 +35,120 @@ you can run them locally and avoid CI failures:
+_I suspect this is not the right place for this documtation, but I think it
+would be good if it was discoverable and "all" developers will see it here_
+
+The test_logger, located in <rr>/common/logging/ can be used to create a `Logger` that by
```
```suggestion
The test_logger, located in `common/logging/` can be used to create a `Logger` that by
```

winksaville replied:
Done

michaelsproul commented on book/src/setup.md:
```diff
@@ -36,6 +35,120 @@ you can run them locally and avoid CI failures:
+would be good if it was discoverable and "all" developers will see it here_
+
+The test_logger, located in <rr>/common/logging/ can be used to create a `Logger` that by
+default returns a NullLogger. But if `--features 'logging/test_logger' is passed while
```
```suggestion
default returns a NullLogger. But if `--features 'logging/test_logger'` is passed while
```

winksaville replied:
Done

michaelsproul commented on common/logging/tests/test-feature-test_logger/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[workspace]
```
Nesting workspaces is considered bad practice, and leads to duplication/maintenance burden in the form of a second `Cargo.lock`.

I would be happy to continue without tests for the `test_logger` feature, because this quickly gets into the weeds of Cargo's build system for minimal benefit. See e.g. https://github.com/rust-lang/cargo/issues/2911 for background. As devs using Lighthouse everyday we will notice if we do something to break the `test_logger`, and a breakage would not be severe (some extra log output for tests only).

Alternatively, you might be able to wire the `test-feature-test_logger` crate into the top-level Cargo.toml to avoid the workspace duplication. If that works, I'm also happy to keep the test.

winksaville replied:
I think more tests are better especially tests of tests, then you know you're being "thorough" or "pedantic" :)

I've moved `/common/logging/tests/test-feature-test_logger` to `/testing/test-test_logger` so now it's not a nested workspace and changed `/common/logging/tests/test.rs` appropriately. Let me know if that pasts muster!

michaelsproul replied:
Ok cool, that looks good. I was overthinking it trying to avoid any `cd {} && cargo test`, but I guess that's exactly what we want to test and hopefully it won't be too fragile :crossed_fingers:

michaelsproul commented on common/logging/tests/test.rs:
```diff
@@ -0,0 +1,53 @@
+            "cd {} && cargo test --features logging/test_logger",
+            test_dir.to_str().unwrap()
+        );
+        //println!("cmd_line: {}", cmd_line);
```
These commented out lines should be deleted before merging

winksaville replied:
Done!

michaelsproul reviewed (changes requested):
Looks good on the whole!

I just had some feedback about testing and docs

winksaville committed:
Changes from Michael Sproul review

In particular moved nested workspace
  /common/logging/tests/test-feature-test_logger to
  /testing/test-test_logger
And now it's just a member of the lighthouse workspace.

winksaville reviewed (commented):
I've incorporated your changes and rebased the change off of latest unstable. Do you want to to squash this in to 1 change or maybe 3:
 1. Adding test_logger to common/logging
 2. Change /slasher to use test_logger
 3. Change /beacon_node/beacon_chain to use test_logger

If we leave it as is a `git bi-sect`, which I've never used, will fail on the first commit in this set. Obviously you can do a squashing merge. But if you'd like to have the 3 commit change I can use `git rebase -i unstable` to do that if you'd like.

What the preference?

michaelsproul commented:
> Do you want to to squash this in to 1 change or maybe 3:

We always do squash merges of PRs, and I think that's OK in this instance too

michaelsproul commented on testing/test-test_logger/.gitignore:
```diff
@@ -0,0 +1 @@
+target/
```
Cargo won't create a `target` directory here so I think we can omit this .gitignore

winksaville replied:
Done

michaelsproul commented on common/logging/tests/test.rs:
```diff
@@ -0,0 +1,57 @@
+mod tests {
+    use std::env;
+    use std::process::Command;
+    //use std::io::Error;
```
This commented out line should be deleted

winksaville replied:
Done

michaelsproul commented on common/logging/tests/test.rs:
```diff
@@ -0,0 +1,57 @@
+#[cfg(test)]
+mod tests {
```
As this is an integration test (lives in `tests/`) the `cfg(test)` and `mod tests` are both redundant

michaelsproul replied:
I mean, you can just remove the level of nesting and just have `#[test]`

winksaville replied:
Done

michaelsproul reviewed (changes requested):
Awesome, looks good. Just a few more tweaks and I think we'll be ready to go

winksaville commented:
AFK, will make the changes tomorrow, txs for spending the time reviewing!

winksaville committed:
Additional review changes from Michael Sproul.

winksaville committed:
Add some documentation for test_logger.

Added additional information to book/src/setup.md about the need for
cmake, using package parameter with cargo test and test_logger.

winksaville commented:
Please note, Ready for review and this is blocking #2580.

michaelsproul reviewed (approved):
bors r+
